File: 730/solution.java
// DP (bottom-up)

class Solution {
    public int countPalindromicSubsequences(String S) {
        int n = S.length();
        int[][] dp = new int[n][n];
        for(int i = 0; i < n; i++) {
            dp[i][i] = 1;
        }
        
        for(int len = 1; len <= n; len++) {
            for(int i = 0; i < n - len; i++) {
                int j = i + len;
                if(S.charAt(i) == S.charAt(j)) {
                    int left = i + 1, right = j - 1;
                    dp[i][j] = dp[i + 1][j - 1] * 2;
                    while(left <= right && S.charAt(i) != S.charAt(left)) {
                        left++;
                    }
                    while(left <= right && S.charAt(i) != S.charAt(right)) {
                        right--;
                    }
                    
                    if(left == right) {
                        dp[i][j] += 1;
                    } else if(left > right) {
                        dp[i][j] += 2;
                    } else {
                        dp[i][j] -= dp[left + 1][right - 1];
                    }
                } else {
                    dp[i][j] = dp[i + 1][j] + dp[i][j - 1] - dp[i + 1][j - 1];
                }
                
                dp[i][j] = (int)(((long)dp[i][j] + kMod) % kMod);
            }
        }
        
        return dp[0][n - 1];
    }
    
    private long kMod = 1000000007;
}

File: 730/solution.rb
# DP (bottom-up

# @param {String} s
# @return {Integer}
def count_palindromic_subsequences(s)
    n = s.size
    dp = Array.new(n) {Array.new(n, 0)}
    n.times do |i| 
        dp[i][i] = 1
    end
    
    (1..n).each do |len|
        (n - len).times do |i|
            j = i + len
            if s[i] == s[j]
                left, right = i + 1, j - 1
                dp[i][j] = dp[left][right] * 2
                left += 1 while left <= right && s[i] != s[left]
                right -= 1 while left <= right && s[i] != s[right]
                    
                if left == right
                    dp[i][j] += 1
                elsif left > right
                    dp[i][j] += 2
                else
                    dp[i][j] -= dp[left + 1][right - 1]
                end
            else
                dp[i][j] = dp[i + 1][j] + dp[i][j - 1] - dp[i + 1][j - 1]
            end
        end
    end
                
    dp[0][n - 1] % (10**9 + 7)
end
